File: src/services/tools/code/readFile.ts
import { z } from 'zod';
import type { ToolDef } from '../types';
import { isTauri } from '../../../lib/tauri';

async function loadFs() {
  if (!isTauri()) return null;
  try {
    return await import('@tauri-apps/plugin-fs');
  } catch {
    return null;
  }
}

export const readFileTool: ToolDef = {
  name: 'read_file',
  description: 'Read the entire contents of a file at the specified path. Returns the content as a string along with file metadata.',
  category: 'code',
  inputSchema: z.object({
    path: z.string().describe('Absolute path to the file to read.'),
  }),
  execute: async ({ path }) => {
    const mod = await loadFs();
    if (!mod) throw new Error('Filesystem access requires the Tauri desktop app');
    const content = await mod.readTextFile(path);
    return { path, content, length: content.length, lines: content.split('\n').length };
  },
};

File: agent/internal/tool/read_file.go
package tool

import (
	"context"
	"fmt"
	"strings"

	"github.com/xz-ephyr/raw-code/agent/pkg/api"
)

func readFileTool() ToolDef {
	return ToolDef{
		Definition: api.ToolDefinition{
			Name:        "read_file",
			Description: "Read the contents of a file from the project filesystem. Supports offset and limit for reading specific sections of large files.",
			Category:    "code",
			Parameters: map[string]api.ParamDef{
				"path":   {Type: "string", Description: "Absolute or relative file path", Required: true},
				"offset": {Type: "number", Description: "Starting line (1-indexed)", Required: false},
				"limit":  {Type: "number", Description: "Max lines to read", Required: false},
			},
		},
		Handler: func(ctx context.Context, e *Executor, params map[string]any) (any, error) {
			path, _ := params["path"].(string)
			if path == "" {
				return nil, fmt.Errorf("path is required")
			}

			content, err := runCmd(ctx, "cat", expandPath(path))
			if err != nil {
				return nil, fmt.Errorf("read file failed: %w", err)
			}

			lines := strings.Split(content, "\n")
			if offset, ok := params["offset"].(float64); ok && offset > 0 {
				start := int(offset) - 1
				if start >= len(lines) {
					start = len(lines)
				}
				if limit, ok := params["limit"].(float64); ok && limit > 0 {
					end := start + int(limit)
					if end > len(lines) {
						end = len(lines)
					}
					content = strings.Join(lines[start:end], "\n")
				} else {
					content = strings.Join(lines[start:], "\n")
				}
			}

			return map[string]any{
				"path":    path,
				"content": content,
				"size":    len(content),
				"lines":   len(lines),
			}, nil
		},
	}
}
